examples/profiling/primitive_instrumentation.cpp

```diff
@@ -0,0 +1,205 @@
+#include <string>
+#include <utility>
+#include <vector>
+
+///////////////////////////////////////////////////////////////////////////////
+char const* const lra_code = R"(block(
+    //
+    // Logistic regression analysis algorithm
+    //
+    //   x: [30, 2]
+    //   y: [30]
+    define(lra, x, y, alpha,
+        block(
+            define(weights, constant(0.0, shape(x, 1))),     // weights: [2]
+            define(transx, transpose(x)),                    // transx:  [2, 30]
+            define(pred, constant(0.0, shape(x, 0))),
+            define(error, constant(0.0, shape(x, 0))),
+            define(gradient, constant(0.0, shape(x, 1))),
+            define(step, 0),
+            while(
+                step < 10,
+                block(
+                    store(pred, 1.0 / (1.0 + exp(-dot(x, weights)))),
+                    store(error, pred - y),                  // error: [30]
+                    store(gradient, dot(transx, error)),     // gradient: [2]
+                    parallel_block(
+                        store(weights, weights - (alpha * gradient)),
+                        store(step, step + 1)
+                    )
+                )
+            ),
+            weights
+        )
+    ),
+    lra
+))";
+
+// Retrieve performance counter data for selected primitives
+//   primitive_instances:     The primitives for which performance counter data
+//                            is required
+//   locality_id:             The locality the performance counter data is going
+//                            to be queried from
+//   counter_name_last_parts: A vector containing the last part of the
+//                            performance counter names.
+//                            e.g. std::vector{ "count/eval", "count/direct_eval" }
+// NOTE: primitive_instances are not verified
+std::map<std::string, std::vector<std::int64_t>> retrieve_counter_data(
+    std::vector<std::string> const& primitive_instances,
+    hpx::naming::id_type locality_id,
+    std::vector<std::string>
+        counter_name_last_parts)
+{
+    // Return value
+    std::map<std::string, std::vector<std::int64_t>> result;
+
+    // Reuse get_counter_values_array calls
+    //   key: primitive type
+    //   value: vector of performance counter values
+    std::map<std::string, std::vector<std::vector<std::int64_t>>>
+        counter_values_pile;
+
+    // Iterate through all provided primitive instances
+    for (auto const& name : primitive_instances)
+    {
+        // Parse the primitive name
+        auto tags =
+            phylanx::execution_tree::compiler::parse_primitive_name(name);
+
+        // TODO: Ensure counter_name_last_part has at least one entry
+
+        // Performance counter values
+        std::vector<std::vector<std::int64_t>>& counter_values =
+            counter_values_pile[tags.primitive];
+        if (counter_values.empty())
+        {
+            // Iterate through the last parts of performance counter names
```

Comment:
You could pre-allocate the memory for the needed number of entries: `counter_values.reserve(counter_name_last_parts.size())`

Reply:
Fixed.